### ISSUE TITLE
Dry-air mass non-conservation in ACE2 models

### ISSUE THREAD

**Issue by brianhenn** [ai2cm/ace#424]
Documenting here that we're observing dry-air mass non-conservation in some ACE2 inference despite configuring these models with `stepper.corrector.conserve_dry_air = True`. Pictured below is an example from a SHiELD-AMIP 1° model's 82-year inference. There is a drift of about 12 Pa in 82 years. The drift rate is small, some runs drift positive and some negative, and the rate seems fairly consistent over time for a given model. Currently I don't think we know quite where this is coming from. 

![W B Chart 8_14_2024, 3_07_19 PM](https://github.com/user-attachments/assets/c000b8d5-d6cf-4ca0-8ef3-41daa0807a85)
 
@spencerkclark also saw this result in ACE-SOM model inference (below).

![Screenshot 2024-07-29 at 3 27 11 PM](https://github.com/user-attachments/assets/b2f0b0e4-9609-4a05-860b-0086bd7f389f)

It also looks like [some drift is evident in the ERA5 ACE2 inference](https://wandb.ai/ai2cm/ace2/reports/inference-mean-weighted_mean_gen-surface_pressure_due_to_dry_air-24-08-14-15-34-05---Vmlldzo5MDM2NjQw) -- @oliverwm1?

Something perhaps relevant happens [when running "dataset inference" using the two SHiELD-AMIP 1° ensemble members as the target (IC2) and prediction (IC1), respectively](https://wandb.ai/ai2cm/ace/runs/bb1icrnr?nw=69ww1pbj5ws), and also including the training ensemble member (IC1) as the monthly reference data to show up on the annual mean plots. Normally there would be no point in including IC1 as the monthly reference data in this kind of dataset inference run, because it will line up exactly with the prediction as they are both just IC1. And that is true for every variable except dry air mass, which [has a strange looking plot](https://wandb.ai/ai2cm/ace/reports/inference-annual-surface_pressure_due_to_dry_air-24-08-14-15-14-06---Vmlldzo5MDM2NTI4). Since the monthly reference data is computed on a different commit than the dataset inference run, it suggest that something is not the same in the dry-air mass computations for these two workflows despite having the same input data.

**Comment by brianhenn** [comment]
See slack conversation that prompted this issue: https://allenai.slack.com/archives/C052FC4NU0G/p1722270889530889

**Comment by oliverwm1** [comment]
@spencerkclark noted this back in March and in that case found that it depended on the scaling used: https://docs.google.com/presentation/d/1TLP_3qKi8waOQdpZ0wzR_D6AWx9wDyCILONpHX8vfkY/edit#slide=id.g26ccc3dd306_0_0. 

We could try casting to double precision before computing/applying the surface pressure correction—the size of the corrections is very small so I think it's plausible these are rounding errors which for some reason did not accumulate with our prior scaling, but do now.

**Comment by brianhenn** [comment]
Reopening this issue because it appears to be non-trivial again on recent main. 

In 80-year evaluations right after #1095 was merged (7fc6b9f8), the dry air mass drift was about 0.5Pa. However on a recent commit off of main (f333772355b4fe23d6d9bdddc3f1f4b1106653ad) with the same model etc., it was about 8Pa. Still small in the big picture (about 10^-6 / yr relative change), but similar to the magnitudes that originally prompted the issue. 

See report showing this change: https://wandb.ai/ai2cm/ace/reports/Dry-air-mass-non-conservation-increased--Vmlldzo5NDI2Njg5

![W B Chart 9_19_2024, 1_02_32 PM](https://github.com/user-attachments/assets/2b9fb218-87b2-485e-914a-a5af54a034db)

**Comment by oliverwm1** [comment]
Maybe we can try to bit more clever than just casting to double precision. E.g. doing the calculation in terms of a reference surface pressure `p_s - 100000` which is smaller and so slightly closer in magnitude to the correction we're imposing.

As you note, the per-step error is about 1e-4 Pa, very small compared to typical surface pressure of 1e5 Pa.

**Comment by oliverwm1** [comment]
An idea I got from Alistair White at EXCLAIM: instead of doing the correction always based on the input/output of a particular step, we could give the corrector 'state' by storing the dry air mass at the initial time step and then always correcting it back to this value during long rollouts. For training I wouldn't expect this to make a noticeable difference, but it could help for the long inference issues shown above.

**Comment by mcgibbon** [comment]
> An idea I got from Alistair White at EXCLAIM: instead of doing the correction always based on the input/output of a particular step, we could give the corrector 'state' by storing the dry air mass at the initial time step and then always correcting it back to this value during long rollouts. For training I wouldn't expect this to make a noticeable difference, but it could help for the long inference issues shown above.

We could do this if we really need to, but this does fundamentally change the API of the stepper, which currently is a state-free function. For example, we would need to re-initialize the stepper to make a new run, or we would get strange, hard-to-diagnose issues. So it would be beneficial to think carefully about the API to support this. Stateful functions are hard to test and diagnose.

One way we could do this without changing APIs is to add a forcing variable that contains the dry air mass, and have the corrector set the new dry air mass to this value (possibly also raising an exception if the initial step dry air mass differs significantly from this value). We would precompute this forcing variable in the data loader.

**Comment by oliverwm1** [comment]
Yes, I agree some care is required for figuring out the right implementation. But it would be good to find a solution for this issue.
